ios/LayoutReanimation/REAAnimationsManager.m

```diff
@@ -100,99 +105,48 @@ - (void)scheduleCleaning
   });
 }
 
-- (void)findRoot:(UIView *)view roots:(NSMutableSet<NSNumber *> *)roots
+- (UIView *)viewForTag:(NSNumber *)tag
 {
-  UIView *currentView = view;
-  NSNumber *lastToRemoveTag = nil;
-  while (currentView != nil) {
-    ViewState state = [_states[currentView.reactTag] intValue];
-    if (state == Disappearing) {
-      return;
-    }
-    if (state == ToRemove) {
-      lastToRemoveTag = currentView.reactTag;
-    }
-    currentView = currentView.superview;
-  }
-  if (lastToRemoveTag != nil) {
-    [roots addObject:lastToRemoveTag];
+  UIView *view = [_reaUiManager viewForReactTag:tag];
+  if (view == nil) {
+    return [_exitingViews objectForKey:tag];
   }
+  return view;
 }
 
-- (BOOL)dfs:(UIView *)root view:(UIView *)view cands:(NSMutableSet<NSNumber *> *)cands
+- (void)endLayoutAnimnationForTag:(NSNumber *)tag cancelled:(BOOL)cancelled type:(NSString *)type
 {
-  NSNumber *tag = view.reactTag;
-  if (tag == nil) {
-    return true;
-  }
-  if (![cands containsObject:tag] && _states[tag] != nil) {
-    return true;
-  }
-  BOOL cannotStripe = false;
-  NSArray<UIView *> *toRemoveCopy = [view.reactSubviews copy];
-  for (UIView *child in toRemoveCopy) {
-    if (![view isKindOfClass:[RCTTextView class]]) {
-      cannotStripe |= [self dfs:root view:child cands:cands];
-    }
-  }
-  if (!cannotStripe) {
-    if (view.reactSuperview != nil) {
-      [_reaUiManager unregisterView:view];
-    }
-    [_states removeObjectForKey:tag];
-    [_viewForTag removeObjectForKey:tag];
-    [_toRemove removeObject:tag];
+  UIView *view = [_exitingViews objectForKey:tag];
+  if ([type isEqualToString:@("exiting")] && view != nil) {
+    [self endAnimationsRecursive:view];
   }
-  return cannotStripe;
 }
 
-- (void)removeLeftovers
+- (void)endAnimationsRecursive:(UIView *)view
 {
-  NSMutableSet<NSNumber *> *roots = [NSMutableSet new];
-  for (NSNumber *viewTag in _toRemove) {
-    UIView *view = _viewForTag[viewTag];
-    if (view == nil) {
-      view = [_reaUiManager viewForReactTag:viewTag];
-      _viewForTag[viewTag] = view;
-    }
-    [self findRoot:view roots:roots];
-  }
-  for (NSNumber *viewTag in roots) {
-    UIView *view = _viewForTag[viewTag];
-    [self dfs:view view:view cands:_toRemove];
+  NSNumber *tag = [view reactTag];
+
+  // we'll remove this view anyway when exiting from recursion,
+  // no need to remove it in `maybeDropAncestors`
+  [_ancestorsToRemove removeObject:tag];
+
+  for (UIView *child in [[view subviews] copy]) {
```

Comment:
There is a need to copy, since changing the array (via `[view removeFromSuperview]` in the next recursion level) during iteration throws an exception.
I doubt copying here costs us much and I feel it's more readable this way than writing an explicit `for` loop over indices.